plugins/generic/shibboleth/pages/ShibbolethHandler.inc.php

```diff
@@ -0,0 +1,437 @@
+				LOG_ERR,
+				"Shibboleth plugin enabled, but not properly configured; failed to find $emailHeader"
+			);
+			Validation::logout();
+			Validation::redirectLogin();
+			return false;
+		}
+
+		$uin = $_SERVER[$uinHeader];
+		$userEmail = $_SERVER[$emailHeader];
+
+		// The UIN must be set; otherwise login failed.
+		if (is_null($uin)) {
+			Validation::logout();
+			Validation::redirectLogin();
+			return false;
+		}
+
+		// Try to locate the user by UIN.
+		$userDao =& DAORegistry::getDAO('UserDAO');
+		$user =& $userDao->getUserByAuthStr($uin, true);
+		if (isset($user)) {
+			syslog(LOG_INFO, "Shibboleth located returning user $uin");
+		} else {
+			// We use the e-mail as a key.
+			if (empty($userEmail)) {
+				syslog(LOG_ERR, "Shibboleth failed to find UIN $uin and no email given.");
+				Validation::logout();
+				Validation::redirectLogin();
+				return false;
+			}
+			$user =& $userDao->getUserByEmail($userEmail);
+
+			if (isset($user)) {
+				syslog(LOG_INFO, "Shibboleth located returning email $userEmail");
+
+				if ($user->getAuthStr() != "") {
+					syslog(
+						LOG_ERR,
+						"Shibboleth user with email $userEmail already has UID"
+					);
+					Validation::logout();
+					Validation::redirectLogin();
+					return false;
+				} else {
+					$user->setAuthStr($uin);
+					$userDao->updateObject($user);
+				}
+			} else {
+				$user = $this->_registerFromShibboleth();
+			}
+		}
+
+		if (isset($user)) {
+			$this->_checkAdminStatus($user);
+
+			$disabledReason = null;
+			$success = Validation::registerUserSession($user, $disabledReason);
+
+			if (!$success) {
+				// @@@ TODO: present user with disabled reason
+				syslog(
+					LOG_ERR,
+					"Disabled user $uin attempted Shibboleth login" .
+						(is_null($disabledReason) ? "" : ": $disabledReason")
+				);
+				Validation::logout();
+				Validation::redirectLogin();
+				return false;
+			}
+
+			return $this->_redirectAfterLogin($request);
+		}
+
+		return false;
+	}
+
+	/**
+	 * @copydoc ShibbolethHandler::activateUser()
+	 */
+	function signIn($args, $request) {
+		return $this->_shibbolethRedirect($args, $request);
+	}
+
+	/**
+	 * Intercept normal logout; redirect to context home page instead
+	 * of login (which would send back to Shibboleth again).
+	 * 
+	 * @param $args array
+	 * @param $request Request
+	 * @return bool
+	 */
+	function signOut($args, $request) {
+		$context = $this->getTargetContext($request);
+		$router = $request->getRouter();
+
+		Validation::logout();
+
+		$contextPath = is_null($context) ? "" : $context->getPath();
+		$returnUrl = $router->url($request, $contextPath);
+		return $request->redirectUrl($returnUrl);
+	}
+
+	/**
+	 * @copydoc ShibbolethHandler::activateUser()
+	 */
+	function validate($args, $request) {
+		return $this->_shibbolethRedirect($args, $request);
+	}
+
+
+	//
+	// Private helper methods
+	//
+	/**
+	 * Get the Shibboleth plugin object
+	 * 
+	 * @return ShibbolethAuthPlugin
+	 */
+	function &_getPlugin() {
```

Comment:
References on objects aren't needed with PHP5+.

Reply:
More cut-and-paste from the predecessor (or maybe some other plugin)—that can go.